wallettool/src/main/java/org/bitcoinj/wallettool/WalletTool.java

```diff
@@ -884,36 +885,64 @@ private void send(PaymentSession session) {
                 return;   // Error message already printed.
         }
 
+        // Complete the transaction
         try {
             wallet.completeTx(req);  // may throw InsufficientMoneyException.
-            if (offline) {
-                wallet.commitTx(req.tx);
-                return;
-            }
+        } catch (InsufficientMoneyException e) {
+            System.err.println("Insufficient funds: have " + wallet.getBalance().toFriendlyString());
+            System.exit(1);
+        }
+        if (offline) {
+            wallet.commitTx(req.tx);
+            return;
+        }
+
+        // Setup network communication (but not PeerGroup)
+        try {
             setup();
+        } catch (BlockStoreException e) {
+            System.out.println("BlockStoreException: " + e.getMessage());
+            e.printStackTrace();
+            System.exit(1);
+        }
+
+        // Send the payment
+        try {
             // No refund address specified, no user-specified memo field.
-            CompletableFuture<PaymentProtocol.Ack> future = session.sendPayment(ImmutableList.of(req.tx), null, null);
-            if (future == null) {
-                // No payment_url for submission so, broadcast and wait.
+            PaymentProtocol.Ack ack = session.sendPayment(ImmutableList.of(req.tx), null, null).get();
+            wallet.commitTx(req.tx);
+            System.out.println("Memo from server: " + ack.getMemo());
+        } catch (ExecutionException e) {
+            try {
+                throw e.getCause();
+            } catch (PaymentProtocolException.InvalidPaymentRequestURL e1) {
+                System.out.println("Missing/Invalid Payment Request URL, broadcasting transaction with PeerGroup");
                 peerGroup.start();
-                peerGroup.broadcastTransaction(req.tx).future().get();
-            } else {
-                PaymentProtocol.Ack ack = future.get();
-                wallet.commitTx(req.tx);
-                System.out.println("Memo from server: " + ack.getMemo());
+                TransactionBroadcast broadcast = peerGroup.broadcastTransaction(req.tx);
+                try {
+                    // Wait for broadcast to be sent
+                    broadcast.future().get();
```

Comment:
@schildbach Maybe this peer broadcast code should be moved into a separate, `private` method. Let me know if you want me to do that and then force-push.

Reply:
Oops, I meant to make this comment on the _new_ PR...